views.py

```diff
@@ -1,7 +1,150 @@
-from app import app
+import flask
+import time
+from app import app,session,request,conn,db
+from konishi_utils import *
 
+@app.route('/favicon.ico')
+def favicon():
+    return flask.send_from_directory(os.path.join(app.root_path, 'static'),'favicon.ico', mimetype='image/vnd.microsoft.icon')
 
+@app.route('/pictures/<filename>')
+def uploaded_file(filename):
+    return send_from_directory(app.config['UPLOAD_FOLDER'],filename)
 
 @app.route('/')
 def home():
     return 'Hello Konishi and >P!'
+
+@app.route('/register',methods=['POST'])
+def register():
+    #Checks if user is already logged in
+    if  'konishi_session' in session:
+        return flask.make_response('<h1>409 Conflict</h1><br>You are already logged in',409)
+    #Checks if username is already taken
+    username = request.form['username']
+    db.execute("SELECT id FROM users WHERE username=?",(username,))
+    if db.fetchone():
+        return flask.make_response('That username is already taken',400)
+    fullname = request.form['fullname'][:20] #Typically a fullname won't be longer than 20 characters
+    #Check db_utils for more info on email checking
+    email = request.form['email']
+    db.execute("SELECT id FROM users WHERE email=?",(email,))
+    if not valid_email(email) or db.fetchone():
+        return flask.make_response('That email is invalid or taken',400)
+    ip_addr = request.remote_addr #Gets the IP Address of the user
+    user_id = make_hash(ip_addr,get_time(),username) #Makes the user ID by hashing the username, IP Address and timestamp
+    password = request.form['password']
+    if len(password)<8: #Password should be longer than 8 characters
+        return flask.make_response('Password needs to be at least 8 characters long',400)
+    else:
+        password = make_hash(password) #Hashes the password
```

Comment:
I think this here is the strongest argument against rolling our own code where well established and tested solutions exist. Using a general purpose hashing function like SHA256 for hashing passwords is a bad idea because it's relatively fast and hashes can be broken much more easily than when using a dedicated slow hash function like [PBKDF2](https://en.wikipedia.org/wiki/PBKDF2). Additionally, storing plain hashes without [salts](https://en.wikipedia.org/wiki/Salt_(cryptography)) opens them up to cracking really easily with rainbow tables.

Now, we got lucky here because I happened to know this, but I'm not a security expert, and I'm sure there's more security weaknesses that can be found. We shouldn't be having these discussions at all — the good thing to do is to use a well established password storage solution where actual security experts already analyzed all kinds of pitfalls and protected against them. There is no advantage to reinvent a very crude wheel where perfectly good and polished wheels exists.

This extends to all kinds of libraries too, not only security ones. There's no point in writing code ourselves which someone else has already extensively written and tested. We don't need to write classes for the models from scratch and manually bind each property to an SQL table column, there's already ORMs for that. Writing own code is hard in general so we should only do it if there is no better option. Yes, writing our own code which we, as the author, fully understand, is much more fun than reading docs for and learning to use an existing thing. But our priority here is creating a good quality end product, not coding for fun.